### ISSUE TITLE
Provide a way to relabel controllers and their pods

### ISSUE THREAD

**Issue by kargakis** [kubernetes/kubernetes#36897]
Currently, the deployment controller relabels replica sets and pods when it needs to adopt a replica set for a deployment. It would be more convenient to have a replica set endpoint that would force the replica set controller to do all the relabeling (change pod template labels, update old pods, change selector). The implementation should be similar to how the rollback spec works for Deployments.

It would help:
* users that want to change their selectors but have no good way to relabel old pods. We are adding `kubectl set selector` in https://github.com/kubernetes/kubernetes/pull/28949 enabled only for services. Eventually, we could enable it for controllers by using the new endpoint.
* in dropping the pod update permission from the deployment controller and move any logic around pods one level down to replica sets (deployments should operate only on replica sets).

@kubernetes/api-review-team @kubernetes/sig-api-machinery @ncdc @liggitt @bgrant0607

**Comment by kargakis** [comment]
cc @mfojtik

**Comment by bgrant0607** [comment]
@kargakis Sorry, quick comments for now.

We may want to think about a subresource, similar to /scale, that could do this uniformly across workload APIs. Maybe even we could extend /scale to be a generic workload-controller interface, but I haven't thought about that much yet.

**Comment by bgrant0607** [comment]
cc @erictune

**Comment by kargakis** [comment]
@bgrant0607 agreed that this should be similar to /scale, not sure about making it generic.

**Comment by kargakis** [comment]
Use-case: Easier Blue-Green deployments: https://github.com/openshift/origin/issues/11954

**Comment by bgrant0607** [comment]
See also comments here:
https://github.com/kubernetes/kubernetes/issues/36859#issuecomment-262022027

Note that one pattern that's safe and doesn't require complex orchestration when one wants to introduce a new attribute is to modify selectors of existing workload controllers to select pods without the new label key, and then just add the key to new pod templates. This isn't possible with RCs, however, since they don't support the more powerful selectors.

**Comment by MarkRx** [comment]
Would causing a relabel to happen spin down old pods and spin up new pods, or would existing pods be reused?

I don't know if this is by design, but at present it isn't really possible to store pod state in labels as changing  the RC causes new pods to be created. Having pod state is useful for services as it allows them to be able to select pods based on their state.

**Comment by kargakis** [comment]
Changing the RC is not causing new pods to be created, you are probably talking about Deployments / DeploymentConfigs. Changing a RC will create new pods if the RC is scaled up or old pods die. Relabeling would merely relabel all old pods to use the new label(s).

**Comment by MarkRx** [comment]
If the RC selector is changed the system would suddenly see 0 pods and try to start scaling up. Since existing pods cannot have their labels changed in an atomic operation when the RC selector is changed there is possible contention.

**Comment by kargakis** [comment]
> If the RC selector is changed the system would suddenly see 0 pods and try to start scaling up. Since existing pods cannot have their labels changed in an atomic operation when the RC selector is changed there is possible contention.

The RC selector should change as the last part of the "transaction" ie. once all old pods and the rc pod template labels have changed. If any of the latter actions fails, we won't change the selector. I guess once we start using transactions in etcd3, this won't be a problem.

**Comment by smarterclayton** [comment]
Don't assume transactions will help us - there are limitations that may
prevent some of these scenarios from being possible.

On Nov 28, 2016, at 8:58 AM, Michail Kargakis <notifications@github.com>
wrote:

If the RC selector is changed the system would suddenly see 0 pods and try
to start scaling up. Since existing pods cannot have their labels changed
in an atomic operation when the RC selector is changed there is possible
contention.

The RC selector should change as the last part of the "transaction" ie.
once all old pods and the rc pod template labels have changed. If any of
the latter actions fails, we won't change the selector. I guess once we
start using transactions in etcd3, this won't be a problem.

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/36897#issuecomment-263277520>,
or mute the thread

**Comment by smarterclayton** [comment]
@bprashanth because we had discussed pod labels changing to track state (like leader).  I think the atomicity challenge remains even with etcd3, so we should at least consider whether this is something that we might indicate another way.

**Comment by lavalamp** [comment]
I think at this point atomic multi object operations is an explicit
non-goal of the kubernetes API system.

**Comment by davidopp** [comment]
Was that officially decided somewhere? I think there are some reasonable use cases for multi-object transactions, for example the way we do pod binding today is not great because it only updates the pod, not the node, so you can't actually detect conflicts. Even just being able to bump the resource version on the node and the pod atomically, while only actually mutating the NodeName in the pod, would be useful. But I suspect there are other useful scenarios (whether the one here is one of those, I don't know).

We were afraid of the implication for sharding etcd, but we never actually needed to shard etcd (yet)...

**Comment by smarterclayton** [comment]
Doing a transaction under the covers of a particular resource is probably
ok.  Trying to do it across resource types via the public API is probably
not.

If we're going to shard on anything, it's going to be resource types, and
it's happening O(soon) for the API extension work.  I doubt a third party
extension will be allowed to have root access to the etcd cluster, and in
any case it's easier if they have their own storage.

On Wed, Nov 30, 2016 at 4:09 AM, David Oppenheimer <notifications@github.com

**Comment by lavalamp** [comment]
Right. API server federation means the set of resources over which a
transaction is even possible is complex, since different groups may be
stored in different etcds or even different storage systems. Even if we
ultimately do decide to allow a few curated transactions, I'm quite
confident it won't happen in 2017.

**Comment by bgrant0607** [comment]
It's off topic for this issue, so please don't discuss it further in this issue, but I agree with smarterclayton and lavalamp: that we should not support atomic transactions across multiple resources, and have documented that:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture.md

OTOH, I could imagine some operation similar to rollback, which performs the relabeling orchestration server-side.

**Comment by bgrant0607** [comment]
See also #14894

**Comment by janetkuo** [comment]
Deployment controller stops relabeling adopted ReplicaSets and Pods after #61615 got merged. Selectors became immutable in apps/v1 endpoints. Can we close this issue now that it's no longer needed by workloads controllers?

**Comment by bgrant0607** [comment]
I agree we are unlikely to get to this, and permitting it would increase complexity of object management.
